### ISSUE TITLE
Compatibility with React 18?

### ISSUE THREAD

**Issue by nemdo** [reach/reach-ui#916]
Components are not compatible with React 18 as it can be seen in package-lock.json:
```
"peerDependencies": {
    "react": "^16.8.0 || 17.x",
    "react-dom": "^16.8.0 || 17.x"
}
```
Eslint throws warnings, but shows 0 vulnerabilities and I haven't had any issue yet.

Is it safe to use Reach UI with React 18? When will you update the libraries? Thank you in advance!

**Comment by johnnyluu** [comment]
I'm finding that Listbox doesn't work (no response after clicking on the trigger) in my new project using React 18, but the same setup works in React 17.

**Comment by chaance** [comment]
I don't support React 18 just yet. There are a number of things that won't play well with concurrent rendering or the new StrictMode behavior. I'm going to start going through all of that this weekend and take stock of what all needs to be done to get there.

**Comment by tajo** [comment]
It would be nice to just relax the `peerDeps` versions of React first. It seems to be mostly working but some package managers like pnpm are more strict and just error out if there's peerDeps mismatch.

**Comment by jeanpan** [comment]
> I don't support React 18 just yet. There are a number of things that won't play well with concurrent rendering or the new StrictMode behavior. I'm going to start going through all of that this weekend and take stock of what all needs to be done to get there.

@chaance, would you mind to list what all needs to be done here?

**Comment by chaance** [comment]
@jeanpan Main thing is checking everything out in `StrictMode` now that React double executes effects, which will likely spot a few bugs we need to fix. In a few places we are tracking renders in the effect (eg, `usePrevious`), and that's gonna be a problem. Another thing is to ensure that we aren't improperly accessing refs during render to prevent concurrent rendering bugs.

There may be some other things, but running the test suite w/ React 18 and checking the storybook examples would be a good way to find out for sure.

**Comment by hood** [comment]
Any news on this?

**Comment by brennarvo** [comment]
I tried running the new `test:react-18` script inside the repo to see if the tests pass with React 18, but was getting path resolution errors: https://github.com/reach/reach-ui/pull/941#issuecomment-1183910011

**Comment by chaance** [comment]
`test:react-16` and `test:react-18` haven't been properly set up yet. I'll respond here when I have everything working.

**Comment by nemdo** [comment]
4 months have passed since the issue was opened and among the 30 dependencies in my project, @reach/dialog is the only one that is not compatible with react 18.x.x. When will you release the new version approximately?

I would love to help, but I definitely don't have the expertise for that. Thank you for all the amazing work!

**Comment by chaance** [comment]
Please keep in mind that this is a free and open source project, free for you to use, fork, or abandon if it no longer works for you. It is maintained very sparingly at the moment, whenever I have the time. There are no updates at the moment, so do with that what you will.

**Comment by dartess** [comment]
> test:react-16 and test:react-18 haven't been properly set up yet. I'll respond here when I have everything working.

@chaance It looks like I've set it up https://github.com/reach/reach-ui/pull/959

**Comment by dartess** [comment]
> Identifying and fixing issues in packages like utils, popover, portal and descendants

I'm starch check and write tests in https://github.com/reach/reach-ui/pull/960

**Comment by LuisxSullivaN** [comment]
Hey sorry I'm curious, to do all these tests and say that they pass in React 18 what are the changes required? Having React 18 installed as a dependency, change `render` to `createRoot` and perform the tests?

**Comment by ahmedyounes** [comment]
tested on my app React 18 
Warning: Expected server HTML to contain a matching <span> in <div>.

Uncaught Error: Hydration failed because the initial UI does not match what was rendered on the server.

**Comment by Jeff-hive9** [comment]
We are looking at moving to React 18 - Does reach-ui support react 18 now ? Appears to be no doco saying so.

**Comment by atulmy** [comment]
Works when
```
"@types/react": "18.0.29",
"@types/react-dom": "18.0.11"
```

**Comment by HuyAms** [comment]
I plan to migrate to React v18. However, the `peerDependencies: "react": "^16.8.0 || 17.x",` prevents me from doing that. Any walkaround?

**Comment by harsimranb** [comment]
It would be really nice to have React v18 setup for this repo.  Projects dependent on this library are blocked on upgrading, or having to fork and update.

**Comment by kentcdodds** [comment]
Jumping in here to mention that when trying with React 19 there are runtime warnings now as well 😭

```
Warning: Invalid prop `ref` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.
    at Menu (/Users/kentcdodds/code/kentcdodds.com/node_modules/@reach/menu-button/src/reach-menu-button.tsx:42:9)
```
